src/OpenSSL/crypto.py

```diff
@@ -217,11 +217,16 @@ def generate_key(self, type, bits):
                 _raise_current_error()
 
         elif type == TYPE_DSA:
-            dsa = _lib.DSA_generate_parameters(
-                bits, _ffi.NULL, 0, _ffi.NULL, _ffi.NULL, _ffi.NULL, _ffi.NULL)
+            dsa = _lib.DSA_new()
             if dsa == _ffi.NULL:
                 # TODO: This is untested.
                 _raise_current_error()
+            res = _lib.DSA_generate_parameters_ex(
+                dsa, bits, _ffi.NULL, 0, _ffi.NULL, _ffi.NULL, _ffi.NULL
+            )
+            if not res == 1:
+                # TODO: This is untested.
```

Comment:
I tend to prefer the `ffi.gc` version I think?  Seems cleaner to me, although it requires a comment explaining what’s going on there which by itself is a red flag.

I’ll leave the decision to you, just make sure nothing leaks. :)